solidity/contracts/BondedECDSAKeepVendor.sol

```diff
@@ -106,7 +111,7 @@ contract BondedECDSAKeepVendor is Proxy {
     /// process. The function emits an event containing the new implementation
     /// address. It can be called after upgrade time delay period has passed since
     /// upgrade initiation.
-    function completeUpgrade() public {
+    function completeUpgrade() public onlyAdmin {
```

Comment:
Why `onlyAdmin`? I think the previous version was better. We made a promise and we are publicly bound to that promise.

Reply:
I liked the previous one too, but right now when we are calling the initialize function we need to restrict the caller to be the admin in case the implementation wants to get `msg.sender` and set it as the contract owner.

Reply:
I don't love it but indeed, there is a risk. Interested in @tintinweb thoughts here.

Reply:
One option here could be: we let anyone cancel the upgrade if the admin did not complete it after some certain timeout. But this timeout should be quite long to avoid censorship. Still digesting it... 🤔

Reply:
Or explicitly state that implementation's initialization function cannot use `msg.sender` but if needed it should expect the address to be provided as a function parameter, so it could be defined by admin in the initialize data.

Reply:
I was thinking about it but rejected this idea as the only thing that guards us against terrible failure is a comment. Not compiler, not unit test, just a comment.

I think the best option is to have `onlyAdmin` on this function and just change the perspective and make it clear in docs. The whole point is not to commit that _the upgrade will be complete in one day_ but that _we will not upgrade the system without a one day delay and announcing it first_.